pubsub.go

```diff
@@ -133,6 +134,35 @@ type PubSubRouter interface {
 	Leave(topic string)
 }
 
+// PubSubControl is the control interface presented by the pubsub system to routers.
+type PubSubControl interface {
+	// PeersForTopic returns the set of peers in a topic; the boolean indicator will be false
+	// if the topic is unknown.
+	PeersForTopic(topic string) (map[peer.ID]struct{}, bool)
+
+	// SendMsg sends a message to a peer.
+	// It returns nil if the message was successfully enqueued for transmission.
+	// If the message was dropped due to full queues, it returns ErrQueueFull.
+	// If the peer was unknown, it returns ErrUnknownPeer.
+	SendMessage(p peer.ID, msg *RPC) error
+
+	// SeenMessage returns true if the message identified by mid has been seen before
+	// by the pubsub system.
+	SeenMessage(mid string) bool
+
+	// Eval schedules a thunk for evaluation in the pubsub process loop.
+	// Returns false if the evaluation was aborted because the pubsub system has been shutdown.
+	Eval(thunk func()) bool
```

Comment:
This is funky.

* Are we allowed to use the rest of the interface outside of a "thunk"?
* Are we using this to protect gossipsub state? If so, what if we just used a lock?

Reply:
The interface in general is unsafe to use outside of the event loop, this should probably be specified in the documentation; will fix.

Reply:
Got it. I wonder if we should put _that_ part into a separate interface and pass it to the thunk? That should make it clear that that's the only safe way to use it.

Reply:
Does `Eval` _wait_ for the thunk to eval?

Reply:
It waits for the evaluation to start, but it doesn't wait for completion.

Reply:
Marked the unsafe outside the event loop methods as such.

Reply:
I'd still like to consider moving these to a separate interface and passing it into the thunk. I _really_ wish go had some kind of goroutine local storage so we could just make all this automatic but it doesn't.

Reply:
Well, the interface gets attached in `Router.Attach` and can get used without explicitly re-entering with `Eval` -- all callbacks to the router from the pubsub system are inside the event loop and that would be awkward to use.
I think the contract should be "you call the pubsub system inside the event loop; if you have a goroutine that is outside (eg a timer, like in gossipsub), use Eval to re-enter the event loop."
We should probably document this in the `Router` interface.

Reply:
Discussed synchronously: All the router callbacks are _also_ called on the main thread so we'd have to feed this new interface in there as well. TL;DR: not worth it.

Instead, we should:

1. Carefully document this.
2. Maybe call this `InvokeInMainThread` (or something less java-y).

Reply:
I think this is a smell. Alternative design proposal in comments above.